src/NuGet.Clients/NuGet.Tools/NuGetPackage.cs

```diff
@@ -1312,5 +1314,32 @@ public int WriteUserOptions(IStream _, string __)
         }
 
         #endregion IVsPersistSolutionOpts
+
+        protected override void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            base.Dispose(disposing);
+
+            if (disposing)
+            {
+                _mcs?.Dispose();
+                _nuGetPowerShellUsageCollector?.Dispose();
+                _semaphore.Dispose();
+                ProjectRetargetingHandler?.Dispose();
+                ProjectUpgradeHandler?.Dispose();
+            }
+
+            _disposed = true;
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
```

Comment:
Are you sure this is necessary? Given that `NuGetPackage` is sealed and `Dispose(bool)` is override (not `virtual`), this suggests that `AsyncPackage` already implements `IDisposable`. I know you put `if (_disposed) return;` in the code, but I think this `Dispose()` method will cause `Dispose(bool)` to be called twice.

Reply:
@zivkan - I did a bit more research to respond to this comment.

- [`IVsPackage.Close()`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.shell.interop.ivspackage.close?view=visualstudiosdk-2019#Microsoft_VisualStudio_Shell_Interop_IVsPackage_Close) invokes `Dispose()` method on `NuGetPackage`. http://index/?query=Package.close&rightProject=VS(Repo)&file=Package.cs&line=1569
- The inheritance hierarchy of `NuGetPackage` doesn't implement `IDisposable` pattern.
- When IDE is closed, `NuGetPackage.Dispose()` method is invoked. (See screenshot)
- I found an example of `Dispose()` method implementation in `VSSDK-Extensibility-Samples` repo.
https://github.com/microsoft/VSSDK-Extensibility-Samples/blob/346f5b0289e5fb8de639ba96fb10703df06cd22d/Code_Sweep/C%23/VsPackage/VsPkg.cs#L92-L109

![image](https://user-images.githubusercontent.com/52756182/116961213-952fc480-ac57-11eb-8a5c-7c5728bc83dd.png)

Reply:
TIL VS's `Package` class implements `virtual Dispose(bool)` without `IDisposable`. How unexpected.

Reply:
Correct @zivkan. I learnt that while responding to your code review comments. Thanks for taking the time to review this PR.